user_sync/connector/directory_ldap.py

```diff
@@ -269,35 +272,55 @@ def iter_users(self, users_filter):
                 if (last_attribute_name != None):
                     self.logger.warn('No email attribute: %s for dn: %s', last_attribute_name, dn)
                 continue
+
+            source_attributes = {}
             
             user = user_sync.connector.helper.create_blank_user()
             user['email'] = email
+            source_attributes['email'] = email
                 
             username, last_attribute_name = self.user_username_formatter.generate_value(record)
             if (username == None and last_attribute_name != None):
                 self.logger.info('No username attribute: %s for dn: %s', last_attribute_name, dn)    
             user['username'] = username if username != None else email
+            if (username != None):
```

Comment:
I disagree, because having no value results in an assertion when you access the attribute source_attributes['username'].  If you are trying to make it clear that there was no source attribute 'username' then return None for its value, don't force the user into awkward constructions like source_attributes.get('username') just for some attributes and not others.